dlt/extract/extract.py

```diff
@@ -367,14 +377,44 @@ def extract(
         load_id = self.extract_storage.create_load_package(source.discover_schema())
         with Container().injectable_context(
             SourceSchemaInjectableContext(source.schema)
-        ), Container().injectable_context(SourceInjectableContext(source)):
+        ), Container().injectable_context(
+            SourceInjectableContext(source)
+        ), Container().injectable_context(
+            LoadPackageStateInjectableContext(
+                load_id=load_id, storage=self.extract_storage.new_packages
+            )
+        ) as load_package:
             # inject the config section with the current source name
             with inject_section(
                 ConfigSectionContext(
                     sections=(known_sections.SOURCES, source.section, source.name),
                     source_state_key=source.name,
                 )
             ):
+                if self.refresh is not None:
+                    _resources_to_drop = (
+                        list(source.resources.extracted) if self.refresh != "drop_dataset" else []
+                    )
+                    _state, _ = pipeline_state(Container())
+                    new_schema, new_state, drop_info = drop_resources(
+                        source.schema,
+                        _state,
+                        resources=_resources_to_drop,
+                        drop_all=self.refresh == "drop_dataset",
+                        state_only=self.refresh == "drop_data",
+                        state_paths="*" if self.refresh == "drop_dataset" else [],
+                    )
+                    _state.update(new_state)
+                    if drop_info["tables"]:
```

Comment:
yeah this is good. but IMO we should also save tables to truncate if `drop_data` is used. so we have `dropped_tables` and `truncated_tables` in package state. no?

Reply:
True, that would be cleaner